File: week18_7_27/PRG_더맵게_민지.py
from heapq import heapify, heappop, heappush

def solution(scoville, K):
    heapify(scoville)
    
    cnt = 0
    while scoville[0] < K:
        # 만약 재료가 2개 미만으로 남았다면 -1를 리턴하고 break
        if len(scoville) < 2:
            cnt = -1
            break
        
        # 아니라면 두개의 재료를 뽑아서 섞기
        a, b = heappop(scoville), heappop(scoville)
        heappush(scoville, a + b*2)
        cnt += 1
        
    return cnt

File: week18_7_27/PRG_더맵게_민지.java
package week18_7_27;

import java.util.PriorityQueue;

public class PRG_더맵게 {
    public static void main(String[] args) {
        int[] scoville = {1, 2, 3, 9, 10, 12};
        int K = 7;

        // Python에서 heapify 함수를 쓰는것과 같은 원리
        // 메모리 절약을 위해서는 heap을 직접 배열에 구현하여 쓰는게 좋음
        PriorityQueue<Integer> heap = new PriorityQueue<Integer>();
        for (int i=0; i<scoville.length; i++) {
            heap.add(scoville[i]);
        }

        int cnt = 0;
        while (heap.peek() < K) {
            if (scoville.length < 2) {
                cnt = -1;
                break;
            }

            int a = heap.poll();
            int b = heap.poll();
            heap.add(a + b*2);
            cnt++;
        }

        System.out.println(cnt);
    }
}
